src/Common/ThreadPool.cpp

```diff
@@ -447,59 +460,66 @@ void ThreadPoolImpl<Thread>::worker(typename std::list<Thread>::iterator thread_
 
         {
             Stopwatch watch;
-            std::unique_lock lock(mutex);
+            std::unique_lock lock(parent_pool.mutex);
             ProfileEvents::increment(
                 std::is_same_v<Thread, std::thread> ? ProfileEvents::GlobalThreadPoolLockWaitMicroseconds : ProfileEvents::LocalThreadPoolLockWaitMicroseconds,
                 watch.elapsedMicroseconds());
 
+            // thread just started and we need to push in the job which was requested
+            // and we can not take it directly (FIFO + priorities), so just push it to the queue.
+            if (new_job)
+            {
+                ALLOW_ALLOCATIONS_IN_SCOPE;
+                parent_pool.jobs.emplace(std::move(*new_job));
+                new_job = std::nullopt;
```

Comment:
it seems like it works worse that the approach with relocking the same lock twice inside schedule. 

I think it's because this way the task becomes available for other threads only once the new thread will start (which takes time in the scenario i'm looking at...)